### ISSUE TITLE
Webpack uses a lot of memory

### ISSUE THREAD

**Issue by rschamp** [LLK/scratch-www#3290]
Recently we've had issues with builds hanging when adding new views, which puts us in a precarious place.

We're still on an old version of webpack, which could be a contributing factor, but some tests by @thisandagain seem to show that simply upgrading isn't a silver bullet for this issue.

We should try to figure out why the memory usage is so bad and do whatever we need to fix it, hopefully also getting up to date with webpack at the same time?

**Comment by picklesrus** [comment]
For reference, removing this require of the Page component should reproduce the memory problems: https://github.com/LLK/scratch-www/blob/develop/src/views/join/join.jsx#L6

**Comment by benjiwheeler** [comment]
@thisandagain @rschamp @picklesrus Is there anywhere else that we have notes on what has been tried already? E.g., @thisandagain can you summarize what you've tried and what the result was? Is there more than what @rschamp wrote above?

**Comment by benjiwheeler** [comment]
Some misc. notes:

* I'm running node 10.15.0 locally
* webpack in www is 2.7.0

Some stuff I've found on the internet...

re: webpack running out of memory while in `--watch` mode, https://github.com/webpack/webpack/issues/1914 :

> I had this problem with the version 10.. of node but, when I upgraded to the v12.4.0 the error gone.

> My problems appear to have gone away completely after upgrading to Webpack4 

> setting `max_old_space_size` [or  `--max-old-space-size`]

> "scripts": {
>    "build": "cross-env NODE_OPTIONS=--max_old_space_size=8192 webpack"
> }

> {
>  "scripts": {
>    "start": "node --max_old_space_size=4096 node_modules/webpack-dev-server/bin/webpack-dev-server.js --inline --progress --port 3000 --open"
>  }
> }

> You can navigate to http://localhost:8080/webpack-dev-server to see a list of all files in memory.

> disabling plugins, e.g. FaviconsWebpackPlugin

**Comment by benjiwheeler** [comment]
I haven't been able to repro the problem on my machine; compiling without that include line works fine.

**Comment by benjiwheeler** [comment]
I was able to repro the problem on my machine.

I attempted to upgrade node to 12.4.0, webpack to 4.2.0, webpack-dev-middleware to 3.2.0, copy-webpack-plugin to 4.4.0, html-webpack-plugin to 3.2.0, and node-sass to 4.12.0.

However, I was not able to get www to build with this configuration. 

One problem came up in the package enhanced-resolve, and seems related to this: https://github.com/webpack/enhanced-resolve/issues/89#issuecomment-399612279

That error message is 'TypeError: callbacks[i] is not a function'. I manually edited the offending file in an attempt to make it work, but couldn't get it to work.

for future reference, I needed to make these changes to webpack.config.js:

```
@@ -126,6 +126,15 @@ module.exports = {
    node: {
        fs: 'empty'
    },
// START ADDED LINES
    optimization: {
        splitChunks: {
            cacheGroups: {
                commons: {
                    name: 'js/common.bundle.js'
                }
            }
        }
    },
// END ADDED LINES
    plugins: [
        new VersionPlugin({length: 5})
    ].concat(routes

...

                'process.env.STATIC_HOST': '"' + (process.env.STATIC_HOST || 'https://cdn2.scratch.mit.edu') + '"',
                'process.env.SCRATCH_ENV': '"' + (process.env.SCRATCH_ENV || 'development') + '"',
                'process.env.SENTRY_DSN': '"' + (process.env.SENTRY_DSN || '') + '"'
// START REMOVED LINES
            }),
            new webpack.optimize.CommonsChunkPlugin({
                name: 'common',
                filename: 'js/common.bundle.js'
// END REMOVED LINES
            })
        ])
};
```

and my changes to package.json were:
```
@@ -50,8 +50,10 @@
    "lodash.defaults": "4.0.1",
    "newrelic": "1.25.4",
    "react-helmet": "5.2.0",
    {+"scratch-gui": "0.1.0-prerelease.20191112225310",+}
    "scratch-parser": "^5.0.0",
    "scratch-storage": [-"^0.5.1"-]{+"^0.5.1",+}
{+    "webpack-cli": "3.3.4"+}
  },
  "devDependencies": {
    "ajv": "6.4.0",
@@ -70,7 +72,7 @@
    "chromedriver": "76.0.0",
    "classnames": "2.2.5",
    "cookie": "0.2.2",
    "copy-webpack-plugin": [-"0.2.0",-]{+"4.4.0",+}
    "create-react-class": "15.6.2",
    "css-loader": "0.23.1",
    "email-validator": "2.0.4",
@@ -90,7 +92,7 @@
    "git-bundle-sha": "0.0.2",
    "glob": "5.0.15",
    "google-libphonenumber": "3.2.5",
    "html-webpack-plugin": [-"2.22.0",-]{+"3.2.0",+}
    "iso-3166-2": "0.4.0",
    "jest": "^23.6.0",
    "json-loader": "0.5.2",
@@ -107,7 +109,7 @@
    "lodash.uniqby": "4.7.0",
    "minilog": "2.0.8",
    "node-dir": "0.1.16",
    "node-sass": [-"4.6.1",-]{+"4.12.0",+}
    "pako": "0.2.8",
    "po2icu": "0.0.2",
    "postcss-loader": "2.0.10",
@@ -126,7 +128,6 @@
    "redux-mock-store": "^1.2.3",
    "redux-thunk": "2.0.1",
    "sass-loader": "6.0.6",
[-    "scratch-gui": "0.1.0-prerelease.20191112225310",-]
    "scratch-l10n": "latest",
    "selenium-webdriver": "3.6.0",
    "slick-carousel": "1.6.0",
@@ -135,8 +136,8 @@
    "tap": "14.2.0",
    "url-loader": "2.1.0",
    "watch": "0.16.0",
    "webpack": [-"2.7.0",-]{+"4.2.0",+}
    "webpack-dev-middleware": [-"2.0.4",-]{+"3.2.0",+}
    "xhr": "2.2.0"
  },
  "jest": {
```

**Comment by benjiwheeler** [comment]
This may have been resolved by https://github.com/LLK/scratch-www/pull/5036

**Comment by benjiwheeler** [comment]
I confirmed that this appears to be resolved now -- at least, removing the join.jsx line that @picklesrus mentions no longer prevents the build from completing.
